### DISCUSSION TITLE
Question: is it possible to store source code in bytecode?

### DISCUSSION THREAD

**Discussion by e2iplayer** [quickjs-ng/quickjs#215] [Q&A] [answered]
How complex it will be to add to store source of function/class etc to bytecode?
To get valid source code when .toString() is called when code was loaded from bytecode?

Let say. I have js code like this:

```
function test()
{
    alert("ab");
}

print(test.toString())
```

When it will be compile it to bytecode and then run  the output will not be as expected.

**Comment by bnoordhuis** [comment]
Converted to a discussion because that should already Just Work(TM) after commit b478329. If it doesn't, please provide more details.

**Comment by saghul** [comment]
Works here:

```
$ ./build/qjs t.js
function test()
{
    alert("ab");
}
```

**Comment by e2iplayer** [comment]
Thank you for the answer. Probably you didn't understand me.

This is an example that demonstrates the problem:

$ git describe --tags
v0.2.0-39-gba8b80f
$ make
$ cat test.js
```
function test()
{
        print("aaa");
}

print("Start test");
print("Source code of the function: " + test.toString());
print("End test");
```

$ ./qjsc -e ./test.js -o test.c
$ gcc ./test.c ./libqjs.a -lm -ldl -pthread

$ ./a.out
```
Start test
Source code of the function:
End test
```

$ ./qjs ./test.js
```
Start test
Source code of the function: function test()
{
        print("aaa");
}
End test
```

As you can see using previously generated bytecode the source code in not available.
Causes a different result when executing bytecode and javascript raw code.

Probably this is expected result by many to minimize bytecode size. Source code is not saved during bytes code save.
In any case, this causes problems with some scripts checking own source code during runtime, so it will be great to have option to save/load field "JSFunctionBytecode.source" when bytecode is saved.

**Reply by saghul** [reply]
Ah I see. I didn't test with a function that was compiled to bytecode.

**Reply by saghul** [reply]
I can reproduce it. @bnoordhuis

**Reply by bnoordhuis** [reply, marked as answer]
Okay, I've turned it back into an issue: #217

**Comment by chqrlie** [comment]
Stripping the source from applications compiled to C should be an option if no longer the default. Quickjs has the extension "use strip" pseudo directive for this purpose. A command line option should be added too.

**Reply by saghul** [reply]
We removed use strip here 🙈
